### DISCUSSION TITLE
Default value for optional parameters

### DISCUSSION THREAD

**Discussion by and1helicopter** [edjCase/JsonRpc#109] [Ideas]
Hello, I think it is necessary to add a default value for method parameters. And then we can set them to optional parameter

https://github.com/edjCase/JsonRpc/blob/f17eba0a36f4b8cbab1dace181155e177cec0a3b/src/EdjCase.JsonRpc.Router/RpcMethodInfo.cs#L91-L110

Somthing, like that:

```
internal class RpcParameterInfo : IRpcParameterInfo
{
    public string Name { get; }
    public Type RawType { get; }
    public bool IsOptional { get; }
    public object? DefaultValue { get; }

    public RpcParameterInfo(string name, Type rawType, bool isOptional, object? defaultValue)
    {
        this.Name = name;
        this.RawType = rawType;
        this.IsOptional = isOptional;
        this.DefaultValue = defaultValue;
    }

    public static RpcParameterInfo FromParameter(ParameterInfo parameterInfo)
    {
        Type parameterType = parameterInfo.ParameterType;
        var defaultValue = parameterInfo.DefaultValue;
        return new RpcParameterInfo(parameterInfo.Name!, parameterInfo.ParameterType, parameterInfo.IsOptional, defaultValue);
    }
}
```
And then mapping all optional parameters, filling them with the default value 
https://github.com/edjCase/JsonRpc/blob/f17eba0a36f4b8cbab1dace181155e177cec0a3b/src/EdjCase.JsonRpc.Router/Defaults/DefaultRpcInvoker.cs#L273-L293

Like that:

```
parameterList = new RpcParameter[methodParameters.Count];
for(var i = 0; i < methodParameters.Count; i++)
{
    var parameterInfo = methodParameters[i];

    var requestParameter = requestParameters
        .SingleOrDefault(x => RpcUtil.NamesMatch(x.Key.AsSpan(), parameterInfo.Name.AsSpan()));

    if (requestParameter.Key == null && !parameterInfo.IsOptional)
        return false;

    if(requestParameter.Value == null)
    {
        if (parameterInfo.RawType == typeof(bool))
            parameterList[i] = parameterInfo.DefaultValue == null ?
                RpcParameter.Boolean(default) :
                RpcParameter.Boolean((bool)Convert.ChangeType(parameterInfo.DefaultValue, typeof(bool)));
        else if (parameterInfo.RawType == typeof(int))
            parameterList[i] = RpcParameter.Number((int)Convert.ChangeType(parameterInfo.DefaultValue, typeof(int)));
        else if (parameterInfo.RawType == typeof(string))
            parameterList[i] = RpcParameter.String((string)Convert.ChangeType(parameterInfo.DefaultValue, typeof(string)));
        else
            parameterList[i] = RpcParameter.Null();
    }
    else
    {
        parameterList[i] = requestParameter.Value;
    }
}
```
In this case, we can omit optional parameters when requesting

**Comment by Gekctek** [comment]
@and1helicopter 
Just pushed out a change under a beta release, just because there might be more changes, but wanted you to have something that works
`5.1.5-beta.1`
Let me know if this works well and ill push it to a real release

If you are curious, I took a different approach. Instead of using `DefaultValue` object and converting it to a `RpcParameter`, you can just give it the value `Type.Missing` and when the function is called, it will use the default value. Before I was setting it to `null` which would override the default value

**Reply by and1helicopter** [reply]
`DefaultValue` are set correctly. Now the problem occurs if not all parameters are passed.
For example, in the case for the method and request below: the `data` parameter is null
```
public async Task<object> SendNotification(
  DateTime eventTime,
  string channel,
  string type,
  int ttl = 3600,
  string title = null,
  Data data = null) {
  ...
```

```
{
  "params": {
    "eventTime": "2023-01-26T17:05:11.839648200",
    "channel": "http",
    "type": "data",
    "data": {
      "text": "text",
    }
  },
  "jsonrpc": "2.0",
  "method": "sendNotification",
  "id": "1"
}
```

**Reply by and1helicopter** [reply]
I think it's because `requestParameters.Count` less than in `methodParameters.Count`
https://github.com/edjCase/JsonRpc/blob/f17eba0a36f4b8cbab1dace181155e177cec0a3b/src/EdjCase.JsonRpc.Router/Defaults/DefaultRpcInvoker.cs#L273-L276

**Comment by Gekctek** [comment]
@and1helicopter 
Good call that fixes it

Try out 5.1.5-beta.2

**Reply by and1helicopter** [reply]
Thanks, it's working fine now

**Reply by Gekctek** [reply]
I'll right. I'll give it another day or so before I push a release just in case there are some other issues. Let me know. Thanks for finding these issues
